gradio/routes.py

```diff
@@ -409,21 +411,22 @@ async def run_predict(
                         if getattr(block, "stateful", False)
                     }
                 session_state = app.state_holder[body.session_hash]
-                iterators = app.iterators[body.session_hash]
                 # The should_reset set keeps track of the fn_indices
                 # that have been cancelled. When a job is cancelled,
                 # the /reset route will mark the jobs as having been reset.
                 # That way if the cancel job finishes BEFORE the job being cancelled
                 # the job being cancelled will not overwrite the state of the iterator.
-                # In all cases, should_reset will be the empty set the next time
-                # the fn_index is run.
-                app.iterators[body.session_hash]["should_reset"] = set()
+                if fn_index in app.iterators_to_reset[body.session_hash]:
+                    iterators = {}
```

Comment:
The basic problem was that we would reset `app.iterators[body.session_hash]["should_reset"]` to an empty set here, meaning that (the old) line 451 would never take effect. That meant we would continue using the previous iterator instead of setting it to `None`, even if the function was reset